### PR TITLE
Add the function to shuffle a vector with a seed.

### PR DESCRIPTION
**Pull request by ple13**

### CODE REVIEW

**Comment by wfa-reviewable** [comment]
This change is [Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39)

File: src/main/cc/any_sketch/crypto/shuffle.cc [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  ASSIGN_OR_RETURN(std::unique_ptr<UniformPseudorandomGenerator> prng,
+                   CreatePrngFromSeed(seed));
+
+  // Samples random values.
+  ASSIGN_OR_RETURN(
+      std::vector<unsigned char> arr,
+      prng->GeneratePseudorandomBytes(data.size() * sizeof(unsigned __int128)));
+
+  unsigned __int128* rand = (unsigned __int128*)arr.data();
+  for (uint64_t i = data.size() - 1; i >= 1; i--) {
```
Comment (benjaminkreuter) [inline src/main/cc/any_sketch/crypto/shuffle.cc:50]: This "looks right" but why reimplement this?  Random shuffle is already available in both the standard library and the absl containers library.

**Review by ple13 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39)* status: 0 of 10 files reviewed, 1 unresolved discussion (waiting on @benjaminkreuter, @kungfucraig, @renjiezh, and @SanjayVas)

___
*[`src/main/cc/any_sketch/crypto/shuffle.cc` line 50 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39#gh-1446352490:-NnjOHFi7200Dvxh90L8:b-gwn9y9) ([raw file](https://github.com/world-federation-of-advertisers/any-sketch/blob/323fa0f3330f38c9ee6ee69ee89bd7ebb32316fd/src/main/cc/any_sketch/crypto/shuffle.cc#L50)):*
<details><summary><i>Previously, benjaminkreuter wrote…</i></summary><blockquote>

This "looks right" but why reimplement this?  Random shuffle is already available in both the standard library and the absl containers library.
</blockquote></details>

In our protocol, the workers 1 and 2 need to shuffle the shares together. So they need to agree to a common random seed and sample the permutation using that seed.

**Review by renjiezh — CHANGES_REQUESTED** [review]
> Reviewed 6 of 10 files at r1, 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39)* status: all files reviewed, 5 unresolved discussions (waiting on @benjaminkreuter, @kungfucraig, @ple13, and @SanjayVas)

___
*[`src/main/cc/any_sketch/crypto/shuffle.cc` line 47 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39#-NnjR1LM0jHdJHh1Pnc2:-NnjR1LM0jHdJHh1Pnc3:b-donh8p) ([raw file](https://github.com/world-federation-of-advertisers/any-sketch/blob/323fa0f3330f38c9ee6ee69ee89bd7ebb32316fd/src/main/cc/any_sketch/crypto/shuffle.cc#L47)):*
> ```cpp
>   ASSIGN_OR_RETURN(
>       std::vector<unsigned char> arr,
>       prng->GeneratePseudorandomBytes(data.size() * sizeof(unsigned __int128)));
> ```

Could you help me to understand why we need `int128` instead of `size_t`?
___
*[`src/main/cc/any_sketch/crypto/shuffle.cc` line 58 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39#-NnjQwFECzjDQi3CsMI4:-NnjQwFECzjDQi3CsMI5:b-y45kud) ([raw file](https://github.com/world-federation-of-advertisers/any-sketch/blob/323fa0f3330f38c9ee6ee69ee89bd7ebb32316fd/src/main/cc/any_sketch/crypto/shuffle.cc#L58)):*
> ```cpp
>     int index = rand[i] % (i + 1);
>     // Swaps the element at current position with the one at position index.
>     uint64_t temp = data[i];
> ```

what about using `std::swap`?
___
*[`src/test/cc/any_sketch/crypto/shuffle_test.cc` line 52 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39#-NnjRbr_5UwZcpM1vpDF:-NnjRbra0n7UmegqRPCf:b9ksfmx) ([raw file](https://github.com/world-federation-of-advertisers/any-sketch/blob/323fa0f3330f38c9ee6ee69ee89bd7ebb32316fd/src/test/cc/any_sketch/crypto/shuffle_test.cc#L52)):*
> ```cpp
> TEST(ShuffleWithSeed, InputSizeAtLeastTwoInvalidSeedFails) {
>   PrngSeed seed;
>   *seed.mutable_key() = std::string(kBytesPerAes256Key - 1, 'a');
> ```

`seed->mutable_key()`?
___
*[`src/test/cc/any_sketch/crypto/shuffle_test.cc` line 92 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39#-NnjS1GnEBX-KkvRyUly:-NnjS1GnEBX-KkvRyUlz:b-dgkyms) ([raw file](https://github.com/world-federation-of-advertisers/any-sketch/blob/323fa0f3330f38c9ee6ee69ee89bd7ebb32316fd/src/test/cc/any_sketch/crypto/shuffle_test.cc#L92)):*
> ```cpp
>   }
> }
> ```

We could add another test that two identical arrays generate the same shuffled output.

**Review by ple13 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39)* status: all files reviewed, 5 unresolved discussions (waiting on @benjaminkreuter, @kungfucraig, @renjiezh, and @SanjayVas)

___
*[`src/main/cc/any_sketch/crypto/shuffle.cc` line 47 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39#-NnjR1LM0jHdJHh1Pnc2:-NnjXykC9cThsxoGTc2Q:bykr9vu) ([raw file](https://github.com/world-federation-of-advertisers/any-sketch/blob/323fa0f3330f38c9ee6ee69ee89bd7ebb32316fd/src/main/cc/any_sketch/crypto/shuffle.cc#L47)):*
<details><summary><i>Previously, renjiezh wrote…</i></summary><blockquote>

Could you help me to understand why we need `int128` instead of `size_t`?
</blockquote></details>

We want to sample 128 random bits for each swap, that's why we use int128 instead of size_t.
___
*[`src/main/cc/any_sketch/crypto/shuffle.cc` line 58 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39#-NnjQwFECzjDQi3CsMI4:-NnjW5J2B6ISV8_IqXvF:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/any-sketch/blob/323fa0f3330f38c9ee6ee69ee89bd7ebb32316fd/src/main/cc/any_sketch/crypto/shuffle.cc#L58)):*
<details><summary><i>Previously, renjiezh wrote…</i></summary><blockquote>

what about using `std::swap`?
</blockquote></details>

Done.
___
*[`src/test/cc/any_sketch/crypto/shuffle_test.cc` line 52 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39#-NnjRbr_5UwZcpM1vpDF:-NnjXmSE1tMeHnye-6Xf:b3i7zyj) ([raw file](https://github.com/world-federation-of-advertisers/any-sketch/blob/323fa0f3330f38c9ee6ee69ee89bd7ebb32316fd/src/test/cc/any_sketch/crypto/shuffle_test.cc#L52)):*
<details><summary><i>Previously, renjiezh wrote…</i></summary><blockquote>

`seed->mutable_key()`?
</blockquote></details>

The seed is an object. seed.mutable_key() gives us the pointer to key.
___
*[`src/test/cc/any_sketch/crypto/shuffle_test.cc` line 92 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39#-NnjS1GnEBX-KkvRyUly:-NnjXx6IBMU5JZE2ip59:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/any-sketch/blob/323fa0f3330f38c9ee6ee69ee89bd7ebb32316fd/src/test/cc/any_sketch/crypto/shuffle_test.cc#L92)):*
<details><summary><i>Previously, renjiezh wrote…</i></summary><blockquote>

We could add another test that two identical arrays generate the same shuffled output.
</blockquote></details>

Done.

**Review by SanjayVas — CHANGES_REQUESTED** [review]
> Reviewed 6 of 10 files at r1, 2 of 4 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39)* status: all files reviewed, 8 unresolved discussions (waiting on @benjaminkreuter, @kungfucraig, @ple13, and @renjiezh)

___
*[`src/main/cc/any_sketch/crypto/shuffle.h` line 27 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39#-NnjugiqDXs9KAMb5rwY:-NnjugiqDXs9KAMb5rwZ:bnrv09e) ([raw file](https://github.com/world-federation-of-advertisers/any-sketch/blob/71fc64e9f4da19ce978e96e9d43bb53a61172dce/src/main/cc/any_sketch/crypto/shuffle.h#L27)):*
> ```objc
> namespace wfa::measurement::common::crypto {
> 
> using any_sketch::PrngSeed;
> ```

I believe we generally avoid using declarations in public namespaces in header files unless it's meant to be exported by the header. I can't find the exact guidance in the [style guide](https://google.github.io/styleguide/cppguide.html#Namespaces), but given that this is used in exactly one place I don't think it's a burden to reference it by the fully-qualified name.
___
*[`src/main/cc/any_sketch/crypto/shuffle.cc` line 49 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39#-NnjwFIc6_JygrXsfkTS:-NnjwFIc6_JygrXsfkTT:undefined) ([raw file](https://github.com/world-federation-of-advertisers/any-sketch/blob/71fc64e9f4da19ce978e96e9d43bb53a61172dce/src/main/cc/any_sketch/crypto/shuffle.cc#L49)):*
> ```cpp
>       prng->GeneratePseudorandomBytes(data.size() * sizeof(unsigned __int128)));
> 
>   unsigned __int128* rand = (unsigned __int128*)arr.data();
> ```





_Suggestion:_
```cpp
absl::uint128
```
___
*[`src/main/cc/math/open_ssl_uniform_random_generator.cc` line 154 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39#-NnjxFmj42tGb9IiJj8p:-NnjxFmj42tGb9IiJj8q:bxrt93z) ([raw file](https://github.com/world-federation-of-advertisers/any-sketch/blob/71fc64e9f4da19ce978e96e9d43bb53a61172dce/src/main/cc/math/open_ssl_uniform_random_generator.cc#L154)):*
> ```cpp
> absl::StatusOr<std::unique_ptr<UniformPseudorandomGenerator>>
> CreatePrngFromSeed(const PrngSeed &seed) {
>   std::vector<unsigned char> key(seed.key().begin(), seed.key().end());
> ```

Add a TODO to use absl::Cord instead of making vector copies once we can upgrade to Protobuf v23.0+. See https://protobuf.dev/reference/cpp/cpp-generated/#cord

We're currently blocked from doing this until we get a new rules_proto release.

**Review by renjiezh — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39)* status: all files reviewed, 4 unresolved discussions (waiting on @benjaminkreuter, @kungfucraig, and @ple13)

**Review by ple13 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39)* status: all files reviewed, 3 unresolved discussions (waiting on @benjaminkreuter, @kungfucraig, and @SanjayVas)

___
*[`src/main/cc/any_sketch/crypto/shuffle.h` line 27 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39#-NnjugiqDXs9KAMb5rwY:-NnkPCbz4A5IKTO-d-hc:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/any-sketch/blob/71fc64e9f4da19ce978e96e9d43bb53a61172dce/src/main/cc/any_sketch/crypto/shuffle.h#L27)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

I believe we generally avoid using declarations in public namespaces in header files unless it's meant to be exported by the header. I can't find the exact guidance in the [style guide](https://google.github.io/styleguide/cppguide.html#Namespaces), but given that this is used in exactly one place I don't think it's a burden to reference it by the fully-qualified name.
</blockquote></details>

Done.
___
*[`src/main/cc/any_sketch/crypto/shuffle.cc` line 49 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39#-NnjwFIc6_JygrXsfkTS:-NnkQkkJE8yBK-JGsl9O:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/any-sketch/blob/71fc64e9f4da19ce978e96e9d43bb53a61172dce/src/main/cc/any_sketch/crypto/shuffle.cc#L49)):*
> ```cpp
>       prng->GeneratePseudorandomBytes(data.size() * sizeof(unsigned __int128)));
> 
>   unsigned __int128* rand = (unsigned __int128*)arr.data();
> ```

Done.
___
*[`src/main/cc/math/open_ssl_uniform_random_generator.cc` line 154 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39#-NnjxFmj42tGb9IiJj8p:-NnkReUq6TfbghTUTu_b:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/any-sketch/blob/71fc64e9f4da19ce978e96e9d43bb53a61172dce/src/main/cc/math/open_ssl_uniform_random_generator.cc#L154)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

Add a TODO to use absl::Cord instead of making vector copies once we can upgrade to Protobuf v23.0+. See https://protobuf.dev/reference/cpp/cpp-generated/#cord

We're currently blocked from doing this until we get a new rules_proto release.
</blockquote></details>

Done.

**Review by SanjayVas — APPROVED** [review]
> Reviewed 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39)* status: all files reviewed, 2 unresolved discussions (waiting on @benjaminkreuter, @kungfucraig, and @ple13)

___
*[`src/main/cc/any_sketch/crypto/shuffle.cc` line 25 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39#-NnkXpnI2IsGZi_fP7dq:-NnkXpnJDeAGFJHLNvMk:b-4j59ap) ([raw file](https://github.com/world-federation-of-advertisers/any-sketch/blob/71fc64e9f4da19ce978e96e9d43bb53a61172dce/src/main/cc/any_sketch/crypto/shuffle.cc#L25)):*
> ```cpp
> namespace wfa::measurement::common::crypto {
> 
> using math::CreatePrngFromSeed;
> ```

Note that it was okay to do this in the .cc file. My other comment was specifically re: non-internal namespaces in header files.

**Review by kungfucraig — CHANGES_REQUESTED** [review]
> Reviewed 1 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39)* status: all files reviewed, 4 unresolved discussions (waiting on @benjaminkreuter and @ple13)

___
*[`src/main/cc/any_sketch/crypto/shuffle.h` line 27 at r4](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39#-NnkUfVZ8mEHXkapZXZE:-NnkUfVZ8mEHXkapZXZF:b-bm148b) ([raw file](https://github.com/world-federation-of-advertisers/any-sketch/blob/7a14410b3bfd0ffe9815b98d648453d3d1736068/src/main/cc/any_sketch/crypto/shuffle.h#L27)):*
> ```objc
> namespace wfa::measurement::common::crypto {
> 
> absl::Status ShuffleWithSeed(std::vector<uint32_t>& data,
> ```

SecureShuffleWithSeed?

Maybe move the Fisher-Yates comment to the header file? I think the choice of algorithm becomes part of the interface if we view this as a reference implementation.
___
*[`src/main/cc/any_sketch/crypto/shuffle.cc` line 47 at r4](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39#-NnkdP-TDAdkcIUFId4s:-NnkdP-TDAdkcIUFId4t:b-kp95n0) ([raw file](https://github.com/world-federation-of-advertisers/any-sketch/blob/7a14410b3bfd0ffe9815b98d648453d3d1736068/src/main/cc/any_sketch/crypto/shuffle.cc#L47)):*
> ```cpp
> 
>   absl::uint128* rand = (absl::uint128*)arr.data();
>   for (uint64_t i = data.size() - 1; i >= 1; i--) {
> ```

What is the advantage in counting down vs. counting up. Also the for loop will not run for data[0]. Is this intended, and if so why?

**Review by ple13 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39)* status: 7 of 10 files reviewed, 4 unresolved discussions (waiting on @benjaminkreuter, @kungfucraig, and @SanjayVas)

___
*[`src/main/cc/any_sketch/crypto/shuffle.h` line 27 at r4](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39#-NnkUfVZ8mEHXkapZXZE:-NnlZQIQ78OoleAq8T-K:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/any-sketch/blob/7a14410b3bfd0ffe9815b98d648453d3d1736068/src/main/cc/any_sketch/crypto/shuffle.h#L27)):*
<details><summary><i>Previously, kungfucraig (Craig Wright) wrote…</i></summary><blockquote>

SecureShuffleWithSeed?

Maybe move the Fisher-Yates comment to the header file? I think the choice of algorithm becomes part of the interface if we view this as a reference implementation.
</blockquote></details>

Done.
___
*[`src/main/cc/any_sketch/crypto/shuffle.cc` line 25 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39#-NnkXpnI2IsGZi_fP7dq:-NnlZKSP2vSedvBHJJNy:b-ucfknh) ([raw file](https://github.com/world-federation-of-advertisers/any-sketch/blob/71fc64e9f4da19ce978e96e9d43bb53a61172dce/src/main/cc/any_sketch/crypto/shuffle.cc#L25)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

Note that it was okay to do this in the .cc file. My other comment was specifically re: non-internal namespaces in header files.
</blockquote></details>

Thanks!
___
*[`src/main/cc/any_sketch/crypto/shuffle.cc` line 47 at r4](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39#-NnkdP-TDAdkcIUFId4s:-NnlYnRk-w1JAUXhHJ3Q:b5c0mue) ([raw file](https://github.com/world-federation-of-advertisers/any-sketch/blob/7a14410b3bfd0ffe9815b98d648453d3d1736068/src/main/cc/any_sketch/crypto/shuffle.cc#L47)):*
<details><summary><i>Previously, kungfucraig (Craig Wright) wrote…</i></summary><blockquote>

What is the advantage in counting down vs. counting up. Also the for loop will not run for data[0]. Is this intended, and if so why?
</blockquote></details>

There is no advantage in counting down. If we count up, we will compute index = i + rand[i] % (n - i) to get a random value in [i, n-1], then swap. When counting down we skip data[0] because there's nothing else in the vector to swap with. Anyway, I think counting up will make the code more readable.

**Review by ple13 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39)* status: 7 of 10 files reviewed, 4 unresolved discussions (waiting on @benjaminkreuter, @kungfucraig, and @SanjayVas)

___
*[`src/main/cc/any_sketch/crypto/shuffle.cc` line 50 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39#gh-1446352490:-NnuN_C-1Hqze6KtK9Aw:bep45dm) ([raw file](https://github.com/world-federation-of-advertisers/any-sketch/blob/323fa0f3330f38c9ee6ee69ee89bd7ebb32316fd/src/main/cc/any_sketch/crypto/shuffle.cc#L50)):*
<details><summary><i>Previously, ple13 (Phi) wrote…</i></summary><blockquote>

In our protocol, the workers 1 and 2 need to shuffle the shares together. So they need to agree to a common random seed and sample the permutation using that seed.
</blockquote></details>

Comment has been added to explain why custom implementation is better than std::shuffle. Experiments show that it takes std::shuffle 10 seconds, and SecureShuffleWithSeed 4.5 second for input of size 40 millions.

**Review by kungfucraig — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39)* status: 7 of 10 files reviewed, 4 unresolved discussions (waiting on @benjaminkreuter and @SanjayVas)

___
*[`src/main/cc/any_sketch/crypto/shuffle.cc` line 50 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39#gh-1446352490:-Nnv6n8u-77N-AhIuDi2:bppcidu) ([raw file](https://github.com/world-federation-of-advertisers/any-sketch/blob/323fa0f3330f38c9ee6ee69ee89bd7ebb32316fd/src/main/cc/any_sketch/crypto/shuffle.cc#L50)):*
<details><summary><i>Previously, ple13 (Phi) wrote…</i></summary><blockquote>

Comment has been added to explain why custom implementation is better than std::shuffle. Experiments show that it takes std::shuffle 10 seconds, and SecureShuffleWithSeed 4.5 second for input of size 40 millions.
</blockquote></details>

The note about the implementations varying across the implementations of the standard library is important as well.

**Review by ple13 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39)* status: 7 of 10 files reviewed, 4 unresolved discussions (waiting on @benjaminkreuter, @kungfucraig, and @SanjayVas)

___
*[`src/main/cc/any_sketch/crypto/shuffle.cc` line 50 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39#gh-1446352490:-NnyieKPCW67lCo5d_XM:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/any-sketch/blob/323fa0f3330f38c9ee6ee69ee89bd7ebb32316fd/src/main/cc/any_sketch/crypto/shuffle.cc#L50)):*
<details><summary><i>Previously, kungfucraig (Craig Wright) wrote…</i></summary><blockquote>

The note about the implementations varying across the implementations of the standard library is important as well.
</blockquote></details>

Done.

**Review by benjaminkreuter — COMMENTED** [review]
> Reviewed 1 of 3 files at r5, 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39)* status: 9 of 10 files reviewed, 3 unresolved discussions (waiting on @kungfucraig and @SanjayVas)

**Review by benjaminkreuter — COMMENTED** [review]
:lgtm:

> Reviewed 5 of 10 files at r1, 1 of 4 files at r2, 1 of 4 files at r4, 1 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39#-:-No-V5h0AXtApYJQu9ko:bnfp4nl)* status: all files reviewed, 3 unresolved discussions (waiting on @kungfucraig and @SanjayVas)

**Review by benjaminkreuter — APPROVED** [review]

**Review by stevenwarejones — APPROVED** [review]
> Reviewed 3 of 10 files at r1, 1 of 4 files at r2, 2 of 3 files at r5, 1 of 1 files at r6, 3 of 3 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39)* status: all files reviewed, 3 unresolved discussions (waiting on @kungfucraig and @SanjayVas)

**Review by ple13 — COMMENTED** [review]
> Reviewed 3 of 10 files at r1, 1 of 4 files at r2, 2 of 3 files at r5, 1 of 1 files at r6, 3 of 3 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/any-sketch/39)* status: all files reviewed, 3 unresolved discussions (waiting on @kungfucraig and @SanjayVas)

**Pull request merged**